internal/pipeline/interpolate.go

```diff
@@ -0,0 +1,134 @@
+package pipeline
+
+import (
+	"github.com/buildkite/agent/v3/internal/ordered"
+	"github.com/buildkite/interpolate"
+)
+
+// This file contains helpers for recursively interpolating all the strings in
+// pipeline objects.
+
+// selfInterpolater describes types that can interpolate themselves in-place.
+// They can call interpolate.Interpolate on strings, or
+// interpolate{Slice,Map,OrderedMap,Any} on their other contents, to do this.
+type selfInterpolater interface {
+	interpolate(interpolate.Env) error
+}
+
+// interpolateAny interpolates (almost) anything in-place. It returns the same
+// type it is passed. When passed a string, it returns a new string. Anything
+// it doesn't know how to interpolate is returned unaltered.
+func interpolateAny(env interpolate.Env, o any) (any, error) {
```

Comment:
could we make this generic? ie 
```suggestion
func interpolateAny[O any](env interpolate.Env, o O) (O, error) {
```
it would mean that we wouldn't have to do the casting further down this file - ie [here](https://github.com/buildkite/agent/pull/2180/files#diff-fc2d5994b8045c71c63083e091e94bc3edad2627c9072efd61283637b57973c5R68) we could just have `s[i] = inte`, but i wonder if Go's type system is smart enough to be disentangling all of the type casts/switches etc...

Reply:
> i wonder if Go's type system is smart enough to be disentangling all of the type casts/switches etc...

It is not.

AFAIK there is presently no way to switch on `O` or `o.(type)`. You can do an ordinary type switch on an interface type, like `switch o := any(o).(type)`, and whilst `o` in each case will have the stronger type, the compiler won't see what you're doing when it gets to `return`. If you're lucky and the case is an interface, you can type assert `o.(O)`, but it rejects conversions like `O(o)`.

Reply:
ah, curses. that's a drag.